design/20190708.certificate-request-crd.md

```diff
@@ -174,6 +188,190 @@ type ObjectReference struct {
+
+The Reason field of the Approved/Denied condition should be set to *who* set the
+condition. Who can be interpreted however makes sense to the approver
+implementation. For example, it may include the API group of an approving policy
+controller, or the client agent of a manual request.
+
+The Message field of the Approved/Denied condition should be set to *why* the
+condition is set. Again, why can be interpreted however makes sense to the
+implementation of the approver. For example, the name of the resource that
+approves this request, the violations which caused the request to be denied, or
+the team to who manually approved the request.
+
+When a CertificateRequest has been Denied, it is the responsibility of the
+referenced issuer to then add a Ready condition with the status of "False",
+along with a relevant Reason and Message.
+
+
+##### RBAC
+
+Approved and Denied conditions are set by requesting against the `/status`
+endpoint of the CertificateRequest resource. This is a divergence of the
+certificates.k8s.io CertificateSigningRequest resource that utilises `approval`,
+as CustomResourceDefinitions may [only define `/status` or `/scale`
+sub-resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources).
+Approvers must therefore have permissions to update the status sub-resource of
+CertificateRequests. Approvers missing this permissions will have their request
+rejected by the API server.
+
+Setting the Approved or Denied conditions are restricted by the approver having
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames:
+ # namesapced signers
+ - "<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>"
+ # cluster scoped signers
+ - "<signer-resource-name>.<signer-group>/<signer-name>"
+ # all signers of this resource name
+ - "<signer-resource-name>.<signer-group>/*"
+```
+
+An example ClusterRole that would grant the permissions to set the Approve and
+Denied conditions of CertificateRequests that reference the cluster scoped
+`myissuers` external issuer, in the group `my-example.io`, with the name `myapp`:
+
+```
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: my-example-io-my-issuer-myapp-approver
+rules:
+  - apiGroups: ["cert-manager.io"]
+    resources: ["signers"]
+    verbs: ["approve"]
+    resourceNames: ["myissuers.my-example.io/myapp"]
+```
+
+These permissions are enforced using a
+[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27)
+that is performed in the cert-manager Webhook component when a user attempts to
+set the Approve or Denied condition. The user of the approver is tested along
+with the IssuerRef of the CertificateRequest.
```

Comment:
```suggestion
These permissions are enforced using the
[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27). 

The cert-manager webhook looks at the SubjectAccessReview provided by the user (or, in the above
example, the approver) and checks that they have the correct RBAC as defined above.
```

Reply:
It is the webhook itself that actually performs the SAR request, rather than the user supplying it to the webhook.